scripts/cluster-upgrade.sh

```diff
@@ -232,55 +232,36 @@ upgrade() {
-    KUBELET_VERSION_COUNT=$(KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc get nodes --no-headers -o custom-columns=VERSION:.status.nodeInfo.kubeletVersion | sort -u | wc -l)
-
-    if [ "$OCP_CURRENT_VERSION" == "$TO" ] && [ "$KUBELET_VERSION_COUNT" == "1" ];
-    then
-        log $OCM_NAME "upgrade" "skipping, already on version $TO"
-        teardown $OCM_NAME $CD_NAMESPACE $CD_NAME
-        return
-    fi
-
-    if [ "$OCP_CURRENT_VERSION" != "" ] && [ "$OCP_CURRENT_VERSION" != "$FROM" ] && [ "$KUBELET_VERSION_COUNT" == "1" ];
-    then
-        log $OCM_NAME "upgrade" "skipping, expect version $FROM, found version $OCP_CURRENT_VERSION"
-        return
-    fi
-
-    # is the desired version supported?
-    ALLOWED_UPDATE=$(KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc get clusterversion version -o json | jq ".status.availableUpdates[] | select(.force == false and .version == \"$TO\")")
-    if [ "$ALLOWED_UPDATE" == "" ];
-    then
-        VERSION_CANDIDATES=$(KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc get clusterversion version -o json | jq ".status.availableUpdates | map(.version) | join(\", \")")
-        log $OCM_NAME $LOG_NAME "Cannot upgrade from $FROM, it is not available in upgrade graph."
-        log $OCM_NAME $LOG_NAME "Candidate versions are: $VERSION_CANDIDATES"
-        teardown $OCM_NAME $CD_NAMESPACE $CD_NAME
-        return
-    fi
-
     # is upgrade already progressing?
     DESIRED_VERSION=$(KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc get clusterversion version -o jsonpath='{.spec.desiredUpdate.version}')
 
     if [ "$DESIRED_VERSION" != "$TO" ];
     then
         # desired version isn't set, to setup and patch clusterversion to kick off the upgrade
         setup $OCM_NAME $CD_NAMESPACE $CD_NAME
-
-        log $OCM_NAME "upgrade" "upgrading from $FROM to $TO"
-
         CHANNEL_NAME=$(get_channel $TO)
 
         # https://issues.redhat.com/browse/OSD-3442
         KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc patch clusterversion version --type merge -p "{\"spec\":{\"overrides\": null}}"
+        KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc patch clusterversion version --type merge -p "{\"spec\":{\"channel\": \"$CHANNEL_NAME\"}}"        
+
+        # Initiate upgrade using `oc adm upgrade` and capture output if failure or already installed as from CLI inbuilt logic
+        log $OCM_NAME "upgrade" "Checking for upgrade eligibility"
+        UPGRADE_COMMAND_OUTPUT=$(KUBECONFIG=$TMP_DIR/kubeconfig-${CD_NAMESPACE} oc adm upgrade --to=$TO 2>&1)
+        UPGRADE_COMMAND_STATUS=$?
+
+        if echo "$UPGRADE_COMMAND_OUTPUT" | grep -q "Cluster is already at version" || [ $UPGRADE_COMMAND_STATUS -eq 1 ]
+        then
+            log $OCM_NAME "upgrade" "${UPGRADE_COMMAND_OUTPUT}"
+            teardown $OCM_NAME $CD_NAMESPACE $CD_NAME
+            return
+        fi
```

Comment:
I don't agree with the suggested change. The way it was before conditions on the result of the `grep` command. As suggested it will always succeed because without an operator `[ ... ]` only checks whether what's in it is nonempty.

That said, a less "pipey" way of doing the same thing would be e.g. `[[ "$UPGRADE_COMMAND_OUTPUT" == *"Cluster is already at version"* ]]`

But I agree scraping messages isn't a good look in general if there's any alternative.

Reply:
@jewzaam , @2uasimojo - 
1. Yup, agreed that parsing through the output of `oc` command is not a good approach. But this was to specifically handle the teardown even when the `oc adm upgrade` resulted in rc=0 if the cluster was already at target version.
2. From a point of view, if we handle all use cases based on rc of `oc adm upgrade` then that would suit just fine. Which means, if we could handle "cluster already at version" message event without requiring any additional teardown logic for even a rc=0 then it might be better. Am yet to understand the relevant changes for this if any.

As for the [] braces, I did intend that the if statement only evaluates the boolean condition for the message or any other failure.. Perhaps my shell scripting has got rusty but introducing brackets gives errors for the message.. Will check it in the morning

Reply:
I agree with @2uasimojo in so far as the bracketing suggestion results in a bash syntax failure - I tried it out:

```
if [ echo "$UPGRADE_COMMAND_OUTPUT" | grep -q "Cluster is already at version" ] || [ $UPGRADE_COMMAND_STATUS -eq 1 ]; then echo "In if block"; fi
-bash: [: missing `]'
grep: ]: No such file or directory
```
If I'm interpreting the above correctly, it looks like pipes take precedence over the brackets in bash, so it breaks it. In any event the code as-is seems to behave appropriately when exercising all possible conditions (grep fails, grep succeeds, COMMAND STATUS=0, COMMAND_STATUS=1)

Reply:
@mrbarge - Thanks a lot! Yeah, I faced failure too based on suggested changes. As of now, I will proceed with moving the `log $OCM_NAME "upgrade" "${UPGRADE_COMMAND_OUTPUT}"` statement outside the if block based on suggestion from @jewzaam .. Regarding scraping the message, yeah I agree it isn't a good practice. Though, I think we can keep it on hold till we figure out a better way to handle (or not handle at all?) the teardown for already installed clusters.